File: sdk/csharp/src/Hjtpx.Captcha.Sdk/Exceptions/NetworkException.cs
using System.Net;

namespace Hjtpx.Captcha.Exceptions;

public class NetworkException : CaptchaException
{
    public string? Host { get; }
    public int? Port { get; }
    public string? Endpoint { get; }
    public HttpRequestException? HttpException { get; }

    public NetworkException(string message) : base(message)
    {
    }

    public NetworkException(string message, Exception innerException) : base(message, innerException)
    {
        if (innerException is HttpRequestException httpEx)
        {
            HttpException = httpEx;
            Host = httpEx.Host;
            Port = httpEx.Port;
        }
    }

    public NetworkException(string message, string host, int? port = null, string? endpoint = null)
        : base(message)
    {
        Host = host;
        Port = port;
        Endpoint = endpoint;
    }

    public NetworkException(string message, Exception innerException, string host, int? port = null)
        : base(message, innerException)
    {
        Host = host;
        Port = port;

        if (innerException is HttpRequestException httpEx)
        {
            HttpException = httpEx;
        }
    }

    public static NetworkException ConnectionTimeout(string host, int? timeoutMs = null) =>
        new($"Connection to {host} timed out after {timeoutMs}ms", host, timeoutMs);

    public static NetworkException ConnectionRefused(string host, int? port = null) =>
        new($"Connection to {host}:{port} was refused", host, port);

    public static NetworkException ConnectionReset(string host) =>
        new($"Connection to {host} was reset", host);

    public static NetworkException HostUnreachable(string host) =>
        new($"Host {host} is unreachable", host);

    public static NetworkException DnsResolutionFailed(string host) =>
        new($"DNS resolution failed for host: {host}", host);

    public static NetworkException SslHandshakeFailed(string host) =>
        new($"SSL/TLS handshake failed for host: {host}", host);

    public static NetworkException RequestAborted(string endpoint) =>
        new($"Request to {endpoint} was aborted", endpoint);

    public override string ToString()
    {
        var sb = new System.Text.StringBuilder();
        sb.AppendLine($"Network Exception: {Message}");

        if (!string.IsNullOrEmpty(Host))
        {
            sb.AppendLine($"Host: {Host}");
        }

        if (Port.HasValue)
        {
            sb.AppendLine($"Port: {Port}");
        }

        if (!string.IsNullOrEmpty(Endpoint))
        {
            sb.AppendLine($"Endpoint: {Endpoint}");
        }

        if (HttpException?.StatusCode != null)
        {
            sb.AppendLine($"HTTP Status: {HttpException.StatusCode}");
        }

        return sb.ToString();
    }
}

File: sdk/php/src/Exception/NetworkException.php
<?php

namespace HJTPX\Captcha\Exception;

class NetworkException extends CaptchaException
{
    protected $host;
    protected $port;
    protected $timeout;

    public function __construct(
        string $message = '',
        int $code = 0,
        ?\Throwable $previous = null,
        string $errorCode = null,
        array $context = [],
        string $host = null,
        int $port = null,
        int $timeout = null
    ) {
        parent::__construct($message, $code, $previous, $errorCode, $context);
        $this->host = $host;
        $this->port = $port;
        $this->timeout = $timeout;
    }

    public function getHost(): ?string
    {
        return $this->host;
    }

    public function getPort(): ?int
    {
        return $this->port;
    }

    public function getTimeout(): ?int
    {
        return $this->timeout;
    }

    public static function connectionTimeout(string $host, int $port, int $timeout): self
    {
        return new self(
            "Connection timeout connecting to {$host}:{$port} (timeout: {$timeout}s)",
            0,
            null,
            'CONNECTION_TIMEOUT',
            ['host' => $host, 'port' => $port, 'timeout' => $timeout],
            $host,
            $port,
            $timeout
        );
    }

    public static function connectionRefused(string $host, int $port): self
    {
        return new self(
            "Connection refused to {$host}:{$port}",
            0,
            null,
            'CONNECTION_REFUSED',
            ['host' => $host, 'port' => $port],
            $host,
            $port
        );
    }

    public static function dnsResolutionFailed(string $host): self
    {
        return new self(
            "DNS resolution failed for host: {$host}",
            0,
            null,
            'DNS_RESOLUTION_FAILED',
            ['host' => $host],
            $host
        );
    }
}
